src/unicorn/builder.rs

```diff
@@ -805,6 +805,19 @@ impl ModelBuilder {
         self.reg_flow_ite(rtype.rd(), mul_node);
     }
 
+    fn model_mulw(&mut self, rtype: RType) {
+        let thirtytwo = self.new_const(32);
+        // sign extend each operand
+        let mut left_sext = self.new_sll(self.reg_node(rtype.rs1()), thirtytwo.clone());
+        left_sext = self.new_sra(left_sext, thirtytwo.clone());
+
+        let mut right_sext = self.new_sll(self.reg_node(rtype.rs2()), thirtytwo.clone());
+        right_sext = self.new_sra(right_sext, thirtytwo);
+
+        let mul_node = self.new_mul(left_sext, right_sext);
```

Comment:
I think using the full 64-bit result of the multiplication is not correct. For example the multiplications of (1<31) * (1<31) is expressible in 64-bit, but overflows in 32-bit (i.e. we expect result 0 and not (1<62)). I think just sign-extending the result similarly to the two operands should be correct.